File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/factory/RecordRefValueFactory.java
package ru.citeck.ecos.records3.record.op.atts.service.value.factory;

import ecos.com.fasterxml.jackson210.databind.JsonNode;
import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.RecordsServiceFactory;
import ru.citeck.ecos.records3.*;
import ru.citeck.ecos.records3.record.op.atts.dto.RecordAtts;
import ru.citeck.ecos.records3.record.op.atts.service.value.AttValue;
import ru.citeck.ecos.records3.record.op.atts.service.value.impl.InnerAttValue;
import ru.citeck.ecos.records3.record.op.atts.service.schema.SchemaAtt;
import ru.citeck.ecos.records3.record.op.atts.service.schema.resolver.AttContext;
import ru.citeck.ecos.records3.record.op.atts.service.schema.write.AttSchemaGqlWriter;
import ru.citeck.ecos.records3.record.op.atts.service.schema.write.AttSchemaWriter;

import java.util.*;

public class RecordRefValueFactory implements AttValueFactory<RecordRef> {

    private static final String ATT_ID = "?id";
    private static final String ATT_LOCAL_ID = "?localId";

    private final RecordsService recordsService;
    private final AttSchemaWriter schemaWriter = new AttSchemaGqlWriter();

    public RecordRefValueFactory(RecordsServiceFactory serviceFactory) {
        this.recordsService = serviceFactory.getRecordsServiceV1();
    }

    @Override
    public AttValue getValue(RecordRef value) {
        return new RecordRefValue(value);
    }

    @Override
    public List<Class<? extends RecordRef>> getValueTypes() {
        return Collections.singletonList(RecordRef.class);
    }

    public class RecordRefValue implements AttValue {

        private final RecordRef ref;
        private final InnerAttValue innerAtts;

        RecordRefValue(RecordRef recordRef) {

            SchemaAtt schemaAtt = AttContext.getCurrentSchemaAtt();
            List<SchemaAtt> innerSchema = schemaAtt.getInner();

            ref = recordRef;

            Map<String, String> attsMap = new HashMap<>();

            StringBuilder sb = new StringBuilder();
            for (SchemaAtt inner : innerSchema) {
                String innerName = inner.getName();
                if (!innerName.equals(ATT_ID) && !innerName.equals(ATT_LOCAL_ID)) {
                    schemaWriter.write(inner, sb);
                    attsMap.put(inner.getName(), sb.toString());
                    sb.setLength(0);
                }
            }

            RecordAtts atts;
            if (attsMap.size() > 0) {
                atts = recordsService.getAtts(Collections.singleton(ref), attsMap, true).get(0);
            } else {
                atts = new RecordAtts(ref);
            }
            JsonNode dataNode = atts.getAttributes().getData().asJson();
            dataNode = schemaWriter.unescapeKeys(dataNode);

            innerAtts = new InnerAttValue(dataNode);

            //todo
            if (attsMap.containsKey("?assoc") && !attsMap.containsKey("?str")) {
                atts.set("?str", atts.getStringOrNull("?assoc"));
            }
        }

        @Override
        public RecordRef getId() {
            return ref;
        }

        @Override
        public String getString() {
            return innerAtts.getString();
        }

        @Override
        public String getDisplayName() {
            return innerAtts.getDisplayName();
        }

        @Override
        public Double getDouble() {
            return innerAtts.getDouble();
        }

        @Override
        public Boolean getBool() {
            return innerAtts.getBool();
        }

        @Override
        public Object getJson() {
            return innerAtts.getJson();
        }

        @Override
        public boolean has(@NotNull String name) {
            return innerAtts.has(name);
        }

        @Override
        public Object getAs(@NotNull String type) {
            return innerAtts.getAs(type);
        }

        @Override
        public Object getAtt(@NotNull String name) {
            return innerAtts.getAtt(name);
        }

        public RecordRef getRef() {
            return ref;
        }
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/factory/RecordRefValueFactory.kt
package ru.citeck.ecos.records3.record.op.atts.service.value.factory

import ecos.com.fasterxml.jackson210.databind.JsonNode
import ru.citeck.ecos.records2.RecordRef
import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.record.op.atts.dto.RecordAtts
import ru.citeck.ecos.records3.record.op.atts.service.schema.SchemaAtt
import ru.citeck.ecos.records3.record.op.atts.service.schema.resolver.AttContext
import ru.citeck.ecos.records3.record.op.atts.service.value.AttValue
import ru.citeck.ecos.records3.record.op.atts.service.value.impl.InnerAttValue
import java.util.*

class RecordRefValueFactory(services: RecordsServiceFactory) : AttValueFactory<RecordRef> {

    companion object {
        private val ATT_ID: String = "?id"
        private val ATT_LOCAL_ID: String = "?localId"
        private val ATT_STR: String = "?str"
        private val ATT_ASSOC: String = "?assoc"

        val ATTS_WITHOUT_LOADING = setOf(
            ATT_ID,
            ATT_LOCAL_ID,
            ATT_STR,
            ATT_ASSOC
        )
    }

    private val recordsService = services.recordsServiceV1
    private val schemaWriter = services.attSchemaWriter

    override fun getValue(value: RecordRef): AttValue? {
        return RecordRefValue(value)
    }

    override fun getValueTypes() = listOf(RecordRef::class.java)

    inner class RecordRefValue(private val ref: RecordRef) : AttValue {

        private val innerAtts: InnerAttValue

        init {
            val schemaAtt: SchemaAtt = AttContext.getCurrentSchemaAtt()
            val innerSchema = schemaAtt.inner

            val attsMap: MutableMap<String, String> = HashMap()
            val sb = StringBuilder()
            for (inner in innerSchema) {
                val innerName: String = inner.name
                if (!ATTS_WITHOUT_LOADING.contains(innerName)) {
                    schemaWriter.write(inner, sb)
                    attsMap[inner.name] = sb.toString()
                    sb.setLength(0)
                }
            }
            val atts: RecordAtts
            if (attsMap.isNotEmpty()) {
                atts = recordsService.getAtts(setOf(ref), attsMap, true)[0]
            } else {
                atts = RecordAtts(ref)
            }

            var dataNode: JsonNode = atts.getAtts().getData().asJson()
            dataNode = schemaWriter.unescapeKeys(dataNode)
            innerAtts = InnerAttValue(dataNode)
        }

        override fun getId(): RecordRef {
            return ref
        }

        override fun asText(): String? {
            return ref.toString()
        }

        override fun getDisplayName(): String? {
            return innerAtts.displayName
        }

        override fun asDouble(): Double? {
            return innerAtts.asDouble()
        }

        override fun asBoolean(): Boolean? {
            return innerAtts.asBoolean()
        }

        override fun asJson(): Any? {
            return innerAtts.asJson()
        }

        override fun has(name: String): Boolean {
            return innerAtts.has(name)
        }

        override fun getAs(type: String): Any? {
            return innerAtts.getAs(type)
        }

        override fun getAtt(name: String): Any? {
            return innerAtts.getAtt(name)
        }
    }
}
